tactics/equality.ml

```diff
@@ -1404,11 +1394,12 @@ let inject_at_positions env sigma l2r eq posns tac =
       let sigma, (injbody,resty) = build_injector e_env !evdref t1' (mkVar e) cpath in
       let injfun = mkNamedLambda (make_annot e Sorts.Relevant) t injbody in
       let sigma,congr = Evd.fresh_global env sigma eq.congr in
-      let pf = applist(congr,[t;resty;injfun;t1;t2]) in
+      (* pf : eq t t1 t2 -> eq resty (injfun t1) (injfun t2) *)
+      let pf = mkApp (congr,[|t; resty; injfun; t1; t2|]) in
       let sigma, pf_typ = Typing.type_of env sigma pf in
-      let inj_clause = apply_on_clause (pf,pf_typ) eq_clause in
-      let pf = Clenv.clenv_value_cast_meta inj_clause in
-      let ty = simplify_args env sigma (clenv_type inj_clause) in
+      let pf_typ = Vars.subst1 mkProp (pi3 @@ destProd sigma pf_typ) in
```

Comment:
Isn't this a hack? Why not `lift (-1)` (assuming an anomaly in the code of `lift` if it produces a rel <= 1), or adding an appropriate function `unlift`, or `strengthen`, or whatever?

Reply:
Technically `lift (-1)` is also a hack. I am confident I've seen cases where it doesn't work properly, probably because of other open variables. Both are used in the code base, but I'd only wager about the correctness of the `subst1` variant. Here it shouldn't matter because `pf_typ` shouldn't contain any rel, but better safe than sorry.

Reply:
So, what to conclude? Keeping the hack but adding a comment that this is an "horrible" hack?

Thanks otherwise for the confirmations about the other questions.